### DISCUSSION TITLE
Disconnected websocket - No exception is raised when sending a message!

### DISCUSSION THREAD

**Discussion by yoniker** [encode/starlette#1798] [Q&A]
How can the server detect a situation when the client didn't receive a message sent by the server,
specifically when the client disconnected not gracefully (=without letting the server know when closing the connection)?

To my understanding, this should have raised an exception, but it doesn't.
The first time when Starlette notices that anything is wrong, is when the ping interval times out.

To easily see that, run the following server:

```python
from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.endpoints import HTTPEndpoint
from starlette.routing import Route
from starlette.endpoints import WebSocketEndpoint
from starlette.routing import Route, WebSocketRoute
import json


websockets = dict()


class SayHealthy(HTTPEndpoint):
    async def get(self, request):
        return JSONResponse({'status': 'websockets service is healthy!'})


class RelayMessageToWebsocket(HTTPEndpoint):
    async def post(self,request):
        try:
            receiver_id = request.path_params['user_id']
            data = await request.json()
            websocket = websockets[receiver_id]
            try:
                await websocket.send(json.dumps(data))
            except Exception as e:
                if receiver_id in websockets: del websockets[receiver_id]
                raise e
            print('Success: message was sent!')
            return JSONResponse({'result': 'message_sent'})
        except:
            print('Error: message was not sent')
            return JSONResponse({'result': 'message_not_sent'})


class WebsocketBehavior(WebSocketEndpoint):
    encoding = 'json'

    async def on_connect(self, websocket):
        user_id = websocket.headers.get('user_id')
        print(f'on connect called with user_id={user_id}')
        websockets[user_id] = websocket
        await websocket.accept()

    async def on_receive(self, websocket, data):
        print('on recieve called')
        #await websocket.send_json(data={'Got this':data})
        print(f'Got this from websocket:{data}')

    async def on_disconnect(self, websocket, close_code):
        print('on disconnect called')
        try:
            user_id = websocket.headers.get('user_id')
            del websockets[user_id]
            print(f'deleted {user_id} from websockets dict')
        except:
            pass


routes = [
    Route("/websockets/healthcheck", SayHealthy),
    Route('/websockets/message_user/{user_id}',RelayMessageToWebsocket),
WebSocketRoute("/websockets/register", WebsocketBehavior)
]




app = Starlette(routes=routes)

```

and run it with a higher ping interval:
`import uvicorn uvicorn.run('app:app',host='0.0.0.0', port=5000,ws_ping_interval=300, ws_ping_timeout=300)`

For the client, use a mobile phone(it's probably true for web as well, but I verified that behavior on a mobile phone ), and put the phone in airplane mode while it is connected to the server.
Call the server via the endpoint /websockets/send_message/<user_id>.
The server will happily describe all of the messages as "sent" for 5 minutes.

If this behavior is expected (to my understanding exception should have been raised), how can I check if the message was actually received by the client?

- [X] Initially raised as discussion https://github.com/encode/starlette/discussions/1780

**Comment by yoniker** [comment]
@Kludex Hey, So I'm confused, the corresponding discussion post was left totally unanswered, and yet you move this issue back to discussions.  I mean... I'm worried this will be left in "discussions" untouched for an indefinite amount of time.
If there is anything else I can provide, I'll be more than happy to :D

**Reply by Kludex** [reply]
Well... The idea of discussions is to have less burden on maintainers, so they can check the discussion first to judge if it should be an issue.

Creating the issue because no one replied for 3 days... It's not really cool.

Locking this as duplicated of #1780.
